include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,461 @@
+#include <coroutine>
+#include <exception>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <seastar/core/future.hh>
+
+// async_generator concept is heavily inspired by P2502R2
+// (https://wg21.link/P2502R2), a proposal accepted into C++23. P2502R2
+// introduced std::generator, which provides a synchronous coroutine
+// mechanism for generating ranges. in contrast, async_generator offers
+// asynchronous generation of element sequences.
+namespace seastar::coroutine::experimental {
+
+template<typename Ref, typename Value = void>
+class async_generator;
+
+namespace internal {
+
+template <typename Yielded> class next_awaiter;
+
+template <typename Yielded>
+class async_generator_promise_base : public seastar::task {
+protected:
+    // a glvalue yield expression is passed to co_yield as its operand. and
+    // the object denoted by this expression is guaranteed to live until the
+    // coroutine resumes. we take advantage of this fact by storing only a
+    // pointer to the denoted object in the promise as long as the result of
+    // dereferencing this pointer is convertible to the Ref type.
+    std::add_pointer_t<Yielded> _value = nullptr;
+
+protected:
+    std::exception_ptr _exception;
+    std::coroutine_handle<> _consumer;
+    task* _waiting_task = nullptr;
+
+    /// awaiter returned by the generator when it produces a new element
+    ///
+    /// There are different combinations of expression types passed to
+    /// \c co_yield and \c Ref. In most cases, zero copies are made. Copies
+    /// are only necessary when \c co_yield requires type conversion.
+    ///
+    /// The following table summarizes the number of copies made for different
+    /// scenarios:
+    ///
+    /// | Ref       | co_yield const T& | co_yield T& | co_yield T&& | co_yield U&& |
+    /// | --------- | ----------------- | ----------- | ------------ | ------------ |
+    /// | T         | 0                 | 0           | 0            | 1            |
+    /// | const T&  | 0                 | 0           | 0            | 1            |
+    /// | T&        | ill-formed        | 0           | ill-formed   | ill-formed   |
+    /// | T&&       | ill-formed        | ill-formed  | 0            | 1            |
+    /// | const T&& | ill-formed        | ill-formed  | 0            | 1            |
+    ///
+    /// When no copies are required, \c yield_awaiter is used. Otherwise,
+    /// \c copy_awaiter is used. The latter converts \c U to \c T, and keeps the converted
+    /// value in it.
+    class yield_awaiter final {
+        async_generator_promise_base* _promise;
+        std::coroutine_handle<> _consumer;
+    public:
+        yield_awaiter(async_generator_promise_base* promise,
+                      std::coroutine_handle<> consumer) noexcept
+            : _promise{promise}
+            , _consumer{consumer}
+        {}
+        bool await_ready() const noexcept {
+            return false;
```

Comment:
Why false? No buffering?

Reply:
I see you later have a buffered type.

Reply:
But, I don't see why await_false is correct. If the generator reached co_yield without suspending, a value should be ready.

Reply:
Ah, you want the caller to consume the value before you proceed.